libportal/notification.c

```diff
@@ -60,22 +70,155 @@ ensure_action_invoked_connection (XdpPortal *portal)
+parse_notification (GTask *task)
+{
+  XdpPortal *portal = g_task_get_source_object (task);
+  CallData *data = g_task_get_task_data (task);
+  const char *key;
+  GVariant *value_outer = NULL;
+
+  while (g_variant_iter_next (data->iter, "{&sv}", &key, &value_outer))
+    {
+      g_autoptr(GVariant) value = value_outer;
+
+      /* In version 1 of the interface requests fail if there are unsupported properties */
+      if (portal->notification_interface_version < 2)
+        {
+          const char *supported_properties_v1[] = {
+            "title",
+            "body",
+            "icon",
+            "priority",
+            "default-action",
+            "default-action-target",
+            NULL
+          };
+
+          if (!g_strv_contains (supported_properties_v1, key))
+            continue;
+        }
+
+      g_variant_builder_add (data->builder, "{sv}", key, value);
+    }
+
+  /* Close what was opened before parsing started */
+  g_variant_builder_close (data->builder);
+
+  g_dbus_connection_call_with_unix_fd_list (portal->bus,
+                                            PORTAL_BUS_NAME,
+                                            PORTAL_OBJECT_PATH,
+                                            "org.freedesktop.portal.Notification",
+                                            "AddNotification",
+                                            g_variant_builder_end (data->builder),
+                                            NULL,
+                                            G_DBUS_CALL_FLAGS_NONE,
+                                            -1,
+                                            data->fd_list,
+                                            g_task_get_cancellable (task),
+                                            call_done_cb,
+                                            g_object_ref (task));
+}
 
 static void
-call_done (GObject *source,
-           GAsyncResult *result,
-           gpointer data)
+get_properties_cb (GObject      *source_object,
+                   GAsyncResult *result,
+                   gpointer      user_data)
 {
-  CallDoneData *call_done_data = data;
+  g_autoptr(GTask) task = G_TASK (user_data);
+  XdpPortal *portal = g_task_get_source_object (task);
+  GError *error = NULL;
+  GVariantIter iter;
+  const char *key;
+  g_autoptr(GVariant) ret = NULL;
+  g_autoptr(GVariant) vardict = NULL;
+  GVariant *value = NULL;
+
+  ret = g_dbus_connection_call_finish (G_DBUS_CONNECTION (source_object), result, &error);
+  if (!ret)
+    {
+      g_task_return_error (task, g_steal_pointer (&error));
+      return;
+    }
+
+  vardict = g_variant_get_child_value (ret, 0);
+  g_variant_iter_init (&iter, vardict);
 
-  call_done_data->callback (G_OBJECT (call_done_data->portal), result, call_done_data->data);
+  while (g_variant_iter_loop (&iter, "{&sv}", &key, &value))
+    {
+      if (strcmp (key, "version"))
+        portal->notification_interface_version = g_variant_get_uint32 (value);
+    }
```

Comment:
Looks like a convoluted way to write this:
```c
g_variant_get (ret, "(&a{sv})", &vardict);
g_variant_lookup (vardict, "version", "u", &portal->notification_interface_version);
```

Reply:
I prefer to not loop twice over the vardict, since we have two properties we need ultimately.

Reply:
Vardicts are not lists but more like hashtables. Even then, I don't think the increased complexity justifies some potential performance. Micro-optimizations without measurements are just not a good idea.

Reply:
fine with me

Reply:
Actually  `g_variant_lookup()` does a linear scan https://docs.gtk.org/glib/method.Variant.lookup.html

Reply:
Not ideal but there are literally 2 properties right now and it won't increase to a point where performance will become relevant.